File: src/karaoke/cdgDecoder.ts
// cdgDecoder.ts
// Low-level CDG packet parser

export interface CDGPacket {
  instruction: number;
  data: Uint8Array;
}

export class CDGDecoder {
  static PACKET_SIZE = 24;

  decode(buffer: ArrayBuffer): CDGPacket[] {
    const packets: CDGPacket[] = [];
    const view = new Uint8Array(buffer);

    for (let i = 0; i < view.length; i += CDGDecoder.PACKET_SIZE) {
      if (i + CDGDecoder.PACKET_SIZE > view.length) break;

      const command = view[i];
      const instruction = view[i + 1];

      // CDG packets always have command 0x09
      if (command !== 0x09) continue;

      const data = view.slice(i + 4, i + 20);
      packets.push({ instruction, data });
    }

    return packets;
  }
}

File: src-tauri/src/karaoke/cdg_decoder.rs
// src-tauri/src/karaoke/cdg_decoder.rs

#[derive(Clone, Debug)]
pub struct CDGPacket {
    pub instruction: u8,
    pub data: [u8; 16],
}

pub struct CDGDecoder;

impl CDGDecoder {
    pub fn decode(buffer: &[u8]) -> Vec<CDGPacket> {
        let mut packets = Vec::new();

        let mut i = 0;
        while i + 24 <= buffer.len() {
            let command = buffer[i];
            let instruction = buffer[i + 1];

            if command == 0x09 {
                let mut data = [0u8; 16];
                data.copy_from_slice(&buffer[i + 4..i + 20]);

                packets.push(CDGPacket { instruction, data });
            }

            i += 24;
        }

        packets
    }
}
